prow/spyglass/gcsartifact.go

```diff
@@ -0,0 +1,162 @@
+	return &GCSArtifact{
+		handle:    handle,
+		link:      link,
+		path:      path,
+		sizeLimit: sizeLimit,
+	}
+}
+
+func fieldsFor(a *GCSArtifact) logrus.Fields {
+	return logrus.Fields{
+		"artifact": a.path,
+	}
+}
+
+// Size returns the size of the artifact in GCS
+func (a *GCSArtifact) Size() (int64, error) {
+	attrs, err := a.handle.Attrs(context.Background())
+	if err != nil {
+		return 0, err
+	}
+	return attrs.Size, nil
+}
+
+// JobPath gets the GCS path of the artifact within the current job
+func (a *GCSArtifact) JobPath() string {
+	return a.path
+}
+
+// CanonicalLink gets the GCS web address of the artifact
+func (a *GCSArtifact) CanonicalLink() string {
+	return a.link
+}
+
+// ReadAt reads len(p) bytes from a file in GCS at offset off
+func (a *GCSArtifact) ReadAt(p []byte, off int64) (n int, err error) {
+	var ctx context.Context
+	if a.gzipped() {
+		return 0, viewers.ErrUnsupportedOp
+	}
+	reader, err := a.handle.NewRangeReader(ctx, off, int64(len(p)))
+	defer reader.Close()
+	if err != nil {
+		return 0, err
+	}
+	n, err = reader.Read(p)
+	return n, err
+}
+
+// ReadAtMost reads at most n bytes from a file in GCS. If the file is compressed (gzip) in GCS, n bytes
+// of gzipped content will be downloaded and decompressed into potentially GREATER than n bytes of content.
+func (a *GCSArtifact) ReadAtMost(n int64) ([]byte, error) {
+	reader, err := a.handle.NewRangeReader(context.Background(), 0, n+1)
+	defer reader.Close()
+	if err != nil {
+		return []byte{}, err
+	}
+	var p []byte
+	var e error
+	if a.gzipped() {
+		gReader, err := gzip.NewReader(reader)
+		if err != nil {
+			return []byte{}, err
+		}
+		p, e = ioutil.ReadAll(gReader)
+	} else {
+		p, e = ioutil.ReadAll(reader)
+	}
+	return p, e
+}
+
+// ReadAll will either read the entire file or throw an error if file size is too big
+func (a *GCSArtifact) ReadAll() ([]byte, error) {
+	size, err := a.Size()
+	if err != nil {
+		return []byte{}, err
+	}
+	if size > a.sizeLimit {
+		return []byte{}, viewers.ErrFileTooLarge
+	}
+	reader, err := a.handle.NewReader(context.Background())
+	defer reader.Close()
+	if err != nil {
+		return []byte{}, err
+	}
+	p, err := ioutil.ReadAll(reader)
+	return p, err
+}
+
+// ReadTail reads the last n bytes from a file in GCS
+func (a *GCSArtifact) ReadTail(n int64) ([]byte, error) {
+	if a.gzipped() {
+		return []byte{}, viewers.ErrUnsupportedOp
+	}
+	size, err := a.Size()
+	if err != nil {
+		return []byte{}, err
+	}
+	offset := size - n
+	reader, err := a.handle.NewRangeReader(context.Background(), offset, -1)
+	defer reader.Close()
+	if err != nil {
+		logrus.WithFields(fieldsFor(a)).WithError(err).Error("Error getting range reader to artifact")
+		return []byte{}, err
+	}
+	read, err := ioutil.ReadAll(reader)
+	return read, err
+}
+
+// gzipped returns whether the file is gzip-encoded in GCS
+func (a *GCSArtifact) gzipped() bool {
+	attrs, err := a.handle.Attrs(context.Background())
+	if err != nil {
```

Comment:
IMO it would be better form for the return signature of this function to be `(bool, error)` so the caller can decide how to handle this error